src/code_800B5F0.c

```diff
@@ -0,0 +1,348 @@
+	"\tlsls r0, 3\n"
+	"\tadds r1, r5, 0x4\n"
+	"\tadds r0, r1\n"
+	"\tldr r0, [r0]\n"
+	"\tcmp r0, 0\n"
+	"\tbeq _0800B9EA\n"
+	"\tbl _call_via_r0\n"
+"_0800B9EA:\n"
+	"\tldrh r0, [r4]\n"
+	"\tadds r0, 0x1\n"
+	"\tstrh r0, [r4]\n"
+	"\tldr r1, _0800BA50\n"
+	"\tlsls r0, 16\n"
+	"\tasrs r0, 16\n"
+	"\tmovs r2, 0\n"
+	"\tldrsh r1, [r1, r2]\n"
+	"\tcmp r0, r1\n"
+	"\tbge _0800BA34\n"
+	"\tmovs r3, 0\n"
+	"\tldrsh r0, [r4, r3]\n"
+	"\tlsls r0, 3\n"
+	"\tadds r0, r5\n"
+	"\tmovs r1, 0x2\n"
+	"\tldrsh r0, [r0, r1]\n"
+	"\tldr r7, _0800BA48\n"
+	"\tcmp r0, r6\n"
+	"\tble _0800B9D6\n"
+"_0800BA10:\n"
+	"\tmovs r2, 0\n"
+	"\tldrsh r1, [r7, r2]\n"
+	"\tmov r3, r8\n"
+	"\tmovs r2, 0\n"
+	"\tldrsh r0, [r3, r2]\n"
+	"\tcmp r1, r0\n"
+	"\tbge _0800BA34\n"
+	"\tldr r2, _0800BA58\n"
+	"\tldr r1, _0800BA54\n"
+	"\tmovs r3, 0\n"
+	"\tldrsh r0, [r7, r3]\n"
+	"\tlsls r0, 3\n"
+	"\tadds r0, r1\n"
+	"\tldrh r0, [r0, 0x2]\n"
+	"\tlsls r0, 8\n"
+	"\tmovs r1, 0x28\n"
+	"\torrs r0, r1\n"
+	"\tstrh r0, [r2]\n"
+"_0800BA34:\n"
+	"\tmovs r0, 0x4\n"
+	"\tbl AckInterrupt\n"
+	"\tpop {r3}\n"
+	"\tmov r8, r3\n"
+	"\tpop {r4-r7}\n"
+	"\tpop {r0}\n"
+	"\tbx r0\n"
+	"\t.align 2, 0\n"
+"_0800BA44: .4byte 0x04000006\n"
+"_0800BA48: .4byte gUnknown_203B0AE\n"
+"_0800BA4C: .4byte gIntrCallbacks\n"
+"_0800BA50: .4byte gUnknown_203B0AC\n"
+"_0800BA54: .4byte gUnknown_202D648\n"
+"_0800BA58: .4byte 0x04000004"
+    );
+}
+#endif
+
+void sub_800BA5C(void)
+{
+    if(gUnknown_203B09A != 0)
+    {
+        gUnknown_203B09C++;
+        xxx_update_bg_sound_input();
+    }
+    gUnknown_203B09A = 1;
+    gUnknown_203B09B = 0;
+    gUnknown_203B09C = 0;
+    nullsub_25();
+
+    while(REG_VCOUNT > 159){}
+
+    gUnknown_203B099 = 1;
+    VBlankIntrWait();
+    nullsub_18();
+}
+
+void xxx_update_bg_sound_input(void)
+{
+    if(gUnknown_203B09B != 0){
+        gUnknown_203B09C++;
+    }
+    else {
+        gUnknown_203B09B = 1;
+        gUnknown_203B09A = 0;
+        gUnknown_203B09C = 0;
+        UpdateBGControlRegisters();
+        if(gUnknown_203B099 != 0)
+            UpdateSound();
+        UpdateInput();
+        gUnknown_203B099 = 0;
+        gUnknown_203B0A4 = gUnknown_203B0A0;
+    }
+}
+
+void Timer3Intr(void)
+{
+    if(gIntrCallbacks[4])
+        gIntrCallbacks[4]();
+    AckInterrupt(INTR_FLAG_TIMER3);
+}
+
+// Unused
+u32 sub_800BB34(void)
+{
+    return gUnknown_203B0A0;
+}
+
+void nullsub_178(void)
+{
+}
+
+// Unused
+void sub_800BB44(void)
+{
+    u16 ie_store;
+    u16 dispcnt_store;
+    bool8 interrupt_flag;
+
+    while(REG_KEYINPUT != KEYS_MASK){}
+
+    interrupt_flag = DisableInterrupts();
+
+    while(REG_VCOUNT < 160){}
+
+    dispcnt_store = REG_DISPCNT;
+    REG_DISPCNT = DISPCNT_FORCED_BLANK;
+
+    *(vu16 *)BG_PLTT = RGB_WHITE;
+
+    ie_store = REG_IE;
+    REG_IE = INTR_FLAG_KEYPAD | INTR_FLAG_GAMEPAK;
+
+    REG_KEYCNT = KEY_AND_INTR | R_BUTTON | L_BUTTON | SELECT_BUTTON; // 0x8304
+    REG_IME = 1;
+
+    SoundBiasReset();
+    asm("swi 0x3");
```

Comment:
Add NONMATCHING.

Reply:
Not really a NONMATCHING as this is an intentional asm call to call the interrupt. There isn't a C equivalent for this. We do a similar thing with `Hang`.

Reply:
Gotcha, carry on then.

Reply:
Is there a way to have an inline function that calls these SWI? Mario & Luigi does that.